### PR TITLE
always retry on DNS error response from name server

### LINKED ISSUE #1519: Always fall back to other name servers on DNS error response codes
**Is your feature request related to a problem? Please describe.**

As mentioned in this [recently merged PR](https://github.com/bluejekyll/trust-dns/pull/1513), in using trust-dns-resolver as the DNS resolver for [Fuchsia](http://fuchsia.dev/), we have noticed that there are some authoritative DNS name servers that respond with a `REFUSED` response to indicate not a fatal/terminal query error but that the requested domain is not in the list of hosts the server would respond to queries about.

We've also now seen a number of servers responding with a `FORMERR` response (format error, [response code 1](https://datatracker.ietf.org/doc/html/rfc1035#section-4.1.1)) to queries we produce, and it's not clear why.

**Describe the solution you'd like**

We were thinking the best solution here could be to always fall back to trying other name servers on an error response code, and only if queries to all specified DNS servers fail would we return a final query error, to make trust-dns-resolver more resilient to these server responses. @bluejekyll, you [mentioned that](https://github.com/bluejekyll/trust-dns/pull/1513#issuecomment-868180954) trust-dns may want to become more lax than it currently is with fallback/terminal errors.

Would it be feasible to always try the fallback servers on an error response, and only bail with some final error code if all servers have already been tried? I would be happy to put together a PR for this if you think this is reasonable.

### PR DESCRIPTION
**Pull request by peterthejohnston**
This PR makes the two changes proposed in #1519:

- In the resolver, always try other name servers when we receive a response with an error code. (i.e. only fail the query terminally if we have already tried all of them.)
- Eventually return the error received from the first name server, so in the case of a "legitimate" error, we prefer the response from the highest priority server we queried.

I do have two outstanding questions:
- For some reason, `ResponseCode::BADMODE` does not lead to a retry in this [this test](https://github.com/bluejekyll/trust-dns/compare/main...peterthejohnston:retry-on-errors?expand=1#diff-90f333e308fae6ca92b573fb451e69b70155c72e3f8d1a7f0c836062d0e7b8d1R372). Any idea what's going on here, or whether `ResponseCode::BADMODE` has some kind of special treatment I'm unaware of?
- Where can I add an integration test for the "return first error received" behavior? In [name_server_pool_tests](https://github.com/bluejekyll/trust-dns/blob/main/tests/integration-tests/tests/name_server_pool_tests.rs)? If so, I think I'll need to pull in `RetryDnsHandle` from the trust-dns-proto crate there.

resolves #1519

### CODE REVIEW

**Comment by bluejekyll** [comment]
> For some reason, ResponseCode::BADMODE does not lead to a retry in this this test. Any idea what's going on here, or whether ResponseCode::BADMODE has some kind of special treatment I'm unaware of?

I'm not aware of anything with `BADMODE` specifically, `BADVERS` is awkward because that's the first value of the extended ResponseCode type that has additional bits added from EDNS, so that shouldn't matter here. But, what could be happening here (I'll need to review your code and the tests more closely) is that `BADMODE` requires EDNS since it falls outside the original 4 bit response code and lands in the extended ResponseCode space. So there may be an issue in the tests related to that... not sure.

> Where can I add an integration test for the "return first error received" behavior? In name_server_pool_tests? If so, I think I'll need to pull in RetryDnsHandle from the trust-dns-proto crate there.

I can't think of a reason why moving RetryDnsHandle into `proto` would be a problem. I don't think it will pull in any external dependencies. That's the primary reason for the separation of all of these crates.

**Review by bluejekyll — APPROVED** [review]
This all looks good to me.

File: crates/proto/src/xfer/retry_dns_handle.rs
Comment (djc) [inline crates/proto/src/xfer/retry_dns_handle.rs:102]: Maybe take here instead of cloning?

File: crates/proto/src/xfer/retry_dns_handle.rs
Comment (djc) [inline crates/proto/src/xfer/retry_dns_handle.rs:106]: If you do this after decreasing `self.remaining_attempts` you can do it without cloning.

File: crates/resolver/src/error.rs
Comment (djc) [inline crates/resolver/src/error.rs]: Maybe this `match` should be in a `ResponseCode` method instead? Like `ResponseCode::should_retry()` or something. Feels like this is tightly coupled to the `ResponseCode` variants/definition.

**Review by djc — COMMENTED** [review]
Looks pretty good to me, though I still have some suggestions for improvements.

File: crates/resolver/src/error.rs
Comment (peterthejohnston) [reply crates/resolver/src/error.rs]: Yeah, I see what you mean. It is tightly coupled to the `ResponseCode` variants, but `ResponseCode` is in the proto crate and IIUC is really meant to map directly on to the RFCs (e.g. [1035](https://datatracker.ietf.org/doc/html/rfc1035#section-4.1.1) for the initial response codes), and I think it would be better to leave more "opinionated" stuff, like which response codes should elicit a retry, which is more subjective, in the resolver crate.

Thoughts?

File: crates/proto/src/xfer/retry_dns_handle.rs
Comment (peterthejohnston) [reply crates/proto/src/xfer/retry_dns_handle.rs:106]: good point, thanks!

File: crates/proto/src/xfer/retry_dns_handle.rs
Comment (peterthejohnston) [reply crates/proto/src/xfer/retry_dns_handle.rs:102]: nice, done.

**Comment by peterthejohnston** [comment]
Also, clippy is complaining on the latest commit but it's unrelated to these changes, so maybe that can be resolved in a separate change.

**Comment by peterthejohnston** [comment]
Ok, after writing an integration test and doing some investigation, it appears that I had some misconceptions about retry behavior. I thought that the "fallback" behavior that I'm trying to modify was performed by `RetryDnsHandle`; in fact, `NameServerPool` is responsible for this and cycles through name servers when an error is received [here](https://github.com/bluejekyll/trust-dns/blob/c658222274090901c810742a1026701c1c094416/crates/resolver/src/name_server/name_server_pool.rs#L332), always returning the last/most recent error received of the highest specificity.

So, my changes to `RetryDnsHandle` are actually ill-conceived. The change to "return the error from the highest priority name server" should actually happen in name_server_pool.rs, so I've updated the PR to reflect that. Also, I think this makes more sense in context: when cycling through our pool of name servers, rather than *always* updating the error unless it is a *less* specific error, instead we *only* update the error if it is a *more* specific error, which allows us to return the error
that is
- first, the most specific
- second, from the highest priority name server (i.e. received earliest).

Sorry about the trash, let me know what you think.

**Comment by peterthejohnston** [comment]
Also, you were right @bluejekyll, only the lower 4 bits of the response code [were being set](https://github.com/bluejekyll/trust-dns/blob/main/crates/proto/src/op/header.rs#L177) in the header. This wasn't caught for the other errors that use more than 4 bits because when truncated they ended being retryable errors as well; BADMODE (`19` i.e. `10011b`) turned into NxDOMAIN (`3` i.e. `11b`).

I've removed all the errors that require `EDNS` from that test.

File: crates/resolver/src/error.rs
Comment (djc) [reply crates/resolver/src/error.rs]: My thoughts are that I'd probably prioritize low coupling over exact alignment of the proto crate with what's defined in the RFC, and also that maybe this is a response to the name `should_retry()` in particular; maybe the name could be less opinionated?

But in the end, maybe @bluejekyll should weigh in on this.

**Comment by djc** [comment]
> So, my changes to `RetryDnsHandle` are actually ill-conceived. The change to "return the error from the highest priority name server" should actually happen in name_server_pool.rs, so I've updated the PR to reflect that. Also, I think this makes more sense in context: when cycling through our pool of name servers, rather than always updating the error unless it is a less specific error, instead we only update the error if it is a more specific error, which allows us to return the error
that is

This makes sense to me. 👍

File: crates/resolver/src/error.rs
Comment (peterthejohnston) [reply crates/resolver/src/error.rs]: > maybe this is a response to the name should_retry() in particular

That's a good point, it could be something more like `retryable`.

> I'd probably prioritize low coupling over exact alignment of the proto crate with what's defined in the RFC

That's fair. I'm of the opinion that the semantics of this match have more to do with `ResolveError` than with `ResponseCode` — we special case `NxDomain | NoError` responses that contain no records, etc. I guess the logic that would make sense to be moved into `ResponseCode` is whether a response code is considered an error, but here we consider basically every response code except `NoError` an error that could lead to retry.

So maybe there should be something like this in `ResponseCode`:

```
impl ResponseCode {
    ...
    pub fn is_error(&self) -> bool {
        match self {
            NoError | Unknown(_) => false,
            FormErr | ServFail | NXDomain | NotImp | Refused | YXDomain | YXRRSet | NXRRSet
            | NotAuth | NotZone | BADVERS | BADSIG | BADKEY | BADTIME | BADMODE | BADNAME
            | BADALG | BADTRUNC | BADCOOKIE => true,
        }
    }
}
```

And then `ResolveError::from_response` can call this. What do you think?

**Review by bluejekyll — APPROVED** [review]
This all looks good. I think there are possibly some error codes that we're over-correcting for, like YXRRSET, which I think is only relevant to dynamic DNS if I remember correctly, but since this is the resolver, I think that's fine.

Thanks for catching the area where this change was needed. I haven't looked at this code in depth in a while, so I missed that nuance between the retry_dns_handle and name_server_pool. Nice work!

**Comment by bluejekyll** [comment]
The cleanliness target failing here should be resolved in #1527

File: crates/resolver/src/error.rs
Comment (djc) [reply crates/resolver/src/error.rs]: Fair enough, but if that doesn't materially improve your current change (to make more responses retryable) maybe it should be a different PR.

Comment (peterthejohnston) [reply crates/resolver/src/error.rs]: SGTM, I can follow up with another PR to do this.

**Comment by peterthejohnston** [comment]
> I think there are possibly some error codes that we're over-correcting for, like YXRRSET, which I think is only relevant to dynamic DNS if I remember correctly, but since this is the resolver, I think that's fine.

OK, sounds good. That makes sense.

> Thanks for catching the area where this change was needed. I haven't looked at this code in depth in a while, so I missed that nuance between the retry_dns_handle and name_server_pool. Nice work!

NP, thanks!

**Comment by peterthejohnston** [comment]
I've updated a relevant doc comment and the PR you mentioned @bluejekyll did resolve the cleanliness target, so if this looks good it should be ready to merge.

(I think maybe the low code coverage stat for the diff is a red herring because the test coverage for the diff is added in a different crate—`tests`.)

**Comment by peterthejohnston** [comment]
Also, I was wondering: what is the process/cadence/criteria for putting out a new point release? We were hoping in Fuchsia to be able to pull in this change relatively soon, without having to diverge from upstream. Could I file an issue to tag a v0.20.4 release?

**Comment by bluejekyll** [comment]
I was planning to cut an alpha release of 0.21.0, I was actually waiting for this PR to land for that. Do you need this to be in a 0.20.0 release? That would require a backport from main to the 0.20 branch at this point (there are some API diffs in main that require a 0.21 version bump.

**Comment by peterthejohnston** [comment]
Ah gotcha, I didn't realize that. Could you point me to the API changes between v0.20.x and v0.21.0?

**Comment by bluejekyll** [comment]
@peterthejohnston, I need to review all the PR's that landed, and add a few more, but this is what I had collected so far: https://github.com/bluejekyll/trust-dns/blob/main/CHANGELOG.md#0210-unreleased

**Comment by bluejekyll** [comment]
And thank you for getting this PR in!

**Pull request merged**
